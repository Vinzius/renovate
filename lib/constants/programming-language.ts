// eslint-disable-next-line typescript-enum/no-enum
export enum ProgrammingLanguage {
  Dart = 'dart',
  Docker = 'docker',
  Elixir = 'elixir',
  Golang = 'golang',
  Java = 'java',
  JavaScript = 'js',
  NET = 'dotnet',
  NodeJS = 'node',
  PHP = 'php',
  Protobuf = 'protobuf',
  Python = 'python',
  Ruby = 'ruby',
  Rust = 'rust',
}
