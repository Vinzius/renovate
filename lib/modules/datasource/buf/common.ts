// Api page size limit 100
export const pageSize = 100;

export const homepageUrl = 'https://buf.build/';

export const defaultRegistryUrl = 'https://api.buf.build/';

export const datasource = 'buf';

export const apiPaths = {
  getRepository:
    'buf.alpha.registry.v1alpha1.RepositoryService/GetRepositoryByFullName',
  listRepositoryTags:
    'buf.alpha.registry.v1alpha1.RepositoryTagService/ListRepositoryTags',
};
