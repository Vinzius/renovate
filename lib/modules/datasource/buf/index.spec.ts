import { getPkgReleases } from '..';
import { Fixtures } from '../../../../test/fixtures';
import * as httpMock from '../../../../test/http-mock';
import { EXTERNAL_HOST_ERROR } from '../../../constants/error-messages';
import { apiPaths, datasource, defaultRegistryUrl, pageSize } from './common';

const depName = 'googleapis/googleapis';
const depId = 'c84e502e-8a35-4bdc-aae4-a5dbdf9a685c';
const getRepositoryExpectedBody = `{"fullName":"${depName}"}`;
const listRepositoryTagsExpectedBody = `{"repositoryId":"${depId}","pageSize":${pageSize},"reverse":true}`;

describe('modules/datasource/buf/index', () => {
  describe('getReleases', () => {
    it('throws for error', async () => {
      httpMock
        .scope(defaultRegistryUrl)
        .post(
          `/${apiPaths.getRepository}`,
          (obj) => obj === getRepositoryExpectedBody
        )
        .replyWithError('error');
      await expect(
        getPkgReleases({
          datasource,
          depName,
        })
      ).rejects.toThrow(EXTERNAL_HOST_ERROR);
    });

    it('returns null for 404', async () => {
      httpMock
        .scope(defaultRegistryUrl)
        .post(
          `/${apiPaths.getRepository}`,
          (obj) => obj === getRepositoryExpectedBody
        )
        .reply(404);
      expect(
        await getPkgReleases({
          datasource,
          depName,
        })
      ).toBeNull();
    });

    it('returns null for empty result', async () => {
      httpMock
        .scope(defaultRegistryUrl)
        .post(
          `/${apiPaths.getRepository}`,
          (obj) => obj === getRepositoryExpectedBody
        )
        .reply(200, {});
      expect(
        await getPkgReleases({
          datasource,
          depName,
        })
      ).toBeNull();
    });

    it('returns null for empty 200 OK', async () => {
      httpMock
        .scope(defaultRegistryUrl)
        .post(
          `/${apiPaths.getRepository}`,
          (obj) => obj === getRepositoryExpectedBody
        )
        .reply(200, { versions: [] });
      expect(
        await getPkgReleases({
          datasource,
          depName,
        })
      ).toBeNull();
    });

    it('throws for 5xx', async () => {
      httpMock
        .scope(defaultRegistryUrl)
        .post(
          `/${apiPaths.getRepository}`,
          (obj) => obj === getRepositoryExpectedBody
        )
        .reply(502);
      await expect(
        getPkgReleases({
          datasource,
          depName,
        })
      ).rejects.toThrow(EXTERNAL_HOST_ERROR);
    });

    it('processes real data', async () => {
      httpMock
        .scope(defaultRegistryUrl)
        .post(
          `/${apiPaths.getRepository}`,
          (obj) => obj === getRepositoryExpectedBody
        )
        .reply(200, Fixtures.get('googleapis-repository.json'));
      httpMock
        .scope(defaultRegistryUrl)
        .post(
          `/${apiPaths.listRepositoryTags}`,
          (obj) => obj === listRepositoryTagsExpectedBody
        )
        .reply(200, Fixtures.get('googleapis-tags.json'));
      const res = await getPkgReleases({
        datasource,
        depName,
      });
      expect(res).toMatchSnapshot();
      expect(res?.releases).toHaveLength(2);
    });
  });
});
