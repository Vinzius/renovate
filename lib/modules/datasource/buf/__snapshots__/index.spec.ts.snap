// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/datasource/buf/index getReleases processes real data 1`] = `
{
  "homepage": "https://buf.build/",
  "registryUrl": "https://api.buf.build/",
  "releases": [
    {
      "releaseTimestamp": "2022-09-06T17:15:22.206Z",
      "version": "62f35d8aed1149c291d606d958a7ce32",
    },
    {
      "releaseTimestamp": "2022-09-08T15:02:32.486Z",
      "version": "8d7204855ec14631a499bd7393ce1970",
    },
  ],
}
`;
